provider/consul/consul_catalog.go

```diff
@@ -298,17 +384,44 @@ func (p *CatalogProvider) getNodes(index map[string][]string) ([]catalogUpdate,
 	return nodes, nil
 }
 
+func merge(cs ...<-chan map[string][]string) <-chan map[string][]string {
+	var wg sync.WaitGroup
+	out := make(chan map[string][]string)
+
+	// Start an output goroutine for each input channel in cs.  output
+	// copies values from c to out until c is closed, then calls wg.Done.
+	output := func(c <-chan map[string][]string) {
+		for n := range c {
+			out <- n
+		}
+		wg.Done()
+	}
+	wg.Add(len(cs))
+	for _, c := range cs {
+		go output(c)
+	}
+
+	// Start a goroutine to close out once all the output goroutines are
+	// done.  This must start after the wg.Add call.
+	go func() {
+		wg.Wait()
+		close(out)
+	}()
+	return out
+}
+
 func (p *CatalogProvider) watch(configurationChan chan<- types.ConfigMessage, stop chan bool) error {
 	stopCh := make(chan struct{})
-	serviceCatalog := p.watchServices(stopCh)
+
+	watchCh := merge(p.watchHealthState(stopCh), p.watchCatalogServices(stopCh))
```

Comment:
Rather than doing a merge make the channel a parameter of both watch* and send the same one to both.